src/main/kotlin/net/ccbluex/liquidbounce/utils/inventory/Slots.kt

```diff
@@ -6,81 +6,116 @@ import net.ccbluex.liquidbounce.utils.item.isNothing
 import net.minecraft.block.BlockState
 import net.minecraft.item.Item
 import net.minecraft.item.ItemStack
+import kotlin.collections.filter
 import kotlin.math.abs
 
-object Slots {
-
-    object Hotbar : SlotGroup<HotbarItemSlot>(
-        List(9) { HotbarItemSlot(it) }
-    ) {
-        fun findSlotIndex(item: Item): Int? {
-            return findSlotIndex { it.item == item }
+fun <T : HotbarItemSlot> SlotGroup<T>.findClosestSlot(item: Item): T? =
+    findClosestSlot { it.item === items }
+
+fun <T : HotbarItemSlot> SlotGroup<T>.findClosestSlot(vararg items: Item): T? =
+    findClosestSlot { it.item in items }
+
+/**
+ * Distance order:
+ * current hand -> offhand -> other slots
+ */
+inline fun <T : HotbarItemSlot> SlotGroup<T>.findClosestSlot(predicate: (ItemStack) -> Boolean): T? {
+    return if (mc.player == null) {
+        null
+    } else {
+        val selected = player.inventory.selectedSlot
+        this.filter { predicate(it.itemStack) }.minByOrNull {
+            when {
+                it is OffHandSlot -> Int.MIN_VALUE + 1
+                it.hotbarSlotForServer == selected -> Int.MIN_VALUE
+                else -> abs(selected - it.hotbarSlotForServer)
+            }
         }
+    }
+}
 
-        inline fun findSlotIndex(predicate: (ItemStack) -> Boolean): Int? {
-            return if (mc.player == null) null else find { predicate(it.itemStack) }?.hotbarSlot
-        }
+fun SlotGroup<*>.hasItem(item: Item): Boolean = any { it.itemStack.item === item }
 
-        fun findClosestItem(vararg items: Item): HotbarItemSlot? {
-            return filter { it.itemStack.item in items }
-                .minByOrNull { abs(player.inventory.selectedSlot - it.hotbarSlotForServer) }
-        }
+fun <T : ItemSlot> SlotGroup<T>.findBestToolToMineBlock(
+    blockState: BlockState,
+    ignoreDurability: Boolean = true
+): T? {
```

Comment:
I know this function was there before this PR, but when thinking about it, I figured that this function should be a part of `ModuleAutoTool`.